ocaml/xapi/xapi_guest_agent.ml

```diff
@@ -97,6 +98,45 @@ let networks path (list: string -> string list) =
 		|> List.map (fun (path, prefix) -> find_all_ips path prefix)
 		|> List.concat
 
+(* This function is passed the "device/vif" node, a function it can use to
+ * find the directory listing of sub-nodes and a function to retrieve the value
+ * with the given path.
+ * If "state" of all VIFs are "4", the return value is true
+ * which means the network paths are optimized.
+ * Or else the return value is false.
+ *)
+let network_paths_optimized path (list: string -> string list) (lookup: string -> string option) =
+	try
+		let _ = List.find (fun vif_id ->
+			let vif_state = lookup (extend (extend path vif_id) "state") in
+			match vif_state with
+			| Some vifstate -> vifstate <> "4"
+			| None -> true
+		) (list path) in
+		false
+	with Not_found -> true
```

Comment:
I tried to stop the compute and return false if any vifstate <> 4, so I used List.find but made the code not easy to follow.
Considering the limited count of VIF and to make code clear, I will update to use List.fold_left per your suggestion.
Thanks for your example.